module/purpose/FieldPlayer/src/FieldPlayer.cpp

```diff
@@ -0,0 +1,195 @@
+                                                                  sensors.Hrw,
+                                                                  cfg.ball_threshold,
+                                                                  cfg.equidistant_threshold,
+                                                                  global_config.player_id);
+
+                // Determine if the game is in a penalty situation
+                bool penalty = game_state.mode.value >= GameState::Mode::DIRECT_FREEKICK
+                               && game_state.mode.value <= GameState::Mode::THROW_IN;
+
+                // If sub_mode is 0, the robot must freeze for referee ball repositioning
+                // If sub_mode is 2, the robot must freeze until the referee calls execute
+                if (penalty && (game_state.secondary_state.sub_mode == 0 || game_state.secondary_state.sub_mode == 2)) {
+                    log<DEBUG>("We are in a freeze penalty situation, do nothing.");
+                    return;
+                }
+
+                // Determine if we need to wait for the other team to kick off
+                // If the ball moves, it is in play
+                bool ball_moved   = (field.Hfw * ball->rBWw).norm() > cfg.ball_off_center_threshold;
+                bool kickoff_wait = !ball_moved && !game_state.our_kick_off
+                                    && (game_state.secondary_time - NUClear::clock::now()).count() > 0;
+                // At this point, if a penalty state is in progress, it must be in sub_mode 1,
+                // which is the setup phase where the robot can position to defend or attack.
+                bool allowed_to_attack = !(kickoff_wait || penalty);
+
+                // Attack if we are closest BUT we have to be in a situation where we are allowed to attack, eg not in
+                // penalty set up phase.
+                if (is_closest && allowed_to_attack) {
+                    log<DEBUG>("Attack!");
+                    emit<Task>(std::make_unique<Attack>(ball_pos));
+                    return;
+                }
+
+                // If we are in the best position to attack, but we can't because of the situation, eg penalty
+                // positioning or opponent kickoff, then we should stick to a good spot and be ready to attack
+                if (is_closest && !allowed_to_attack) {
+                    log<DEBUG>("Ready attack!");
+                    emit<Task>(std::make_unique<ReadyAttack>());
+                    return;
+                }
+
+                // If we can't attack, eg another robot is attacking, we don't want to get in the way.
+                // If we are closest to our goals (ignoring the attacking player), then stand back to defend.
+                // If there's no robots, assume we are alone and are the furthest back
+                // Shouldn't happen, as that should make us the attacker
+                bool furthest_back = robots
```

Comment:
Does this run into any issues ever when two robots are close together, like due to noise, an oscillation between two robots being the furthest back?

Reply:
I haven't really encountered issues with oscillating. Getting the kick in will help, but I am trying to keep robots apart to prevent these kinds of issues. I'm sure there are things to improve on this front though, and it might be more obvious on the real robot. When robots are close together, they tend to have lower error.